add a demo version of koptreader in new UI KPV

chrox opened the pull request:
Only page down and page up are working in this demo. More will be added soon.

chrox committed:
add screen dpi property in koptcontext

chrox committed:
add koptreader demo

chrox commented:
What's interesting is that even though the Kpw has a faster CPU than K3 the reflowing process is even slower in Kpw. I will look into this more carefully and find out why this could happen.

Update: No. With more controlled tests the result now favors Kpw which speeds up reflowing by roughly 20%. It's much more reasonable now.

chrox committed:
update dst_width and dst_height in k2settings

houqp commented:
wow, that's awesome :)

too sad that I will have exams today and tomorrow :( 
so I won't be able to test thoroughly until tomorrow night.

houqp commented:
I tried open `test/tall.pdf` and was trapped in the do loop in line 579 in `pdf.c`. Thus it keeps printing `reading page:0,0,0,0 zoom:0.00 dpi:0` in terminal. Can you reproduce it in your environment?

Open `test/2col.pdf` has the same effect.

chrox commented:
No. I have no such problem. I updated libk2pdfopt in this Patch. So have you run `git submodule update` and `make cleanthirdparty thirdparty`?

chrox committed:
recalculate page size after each call of drawPage
Only after renderPage is called can koptdocument update
reflowed page size. And the recalculate acts as a page
size updater here.

chrox committed:
cache page size after page is reflowed
So that when page dimension is recalculated right page size
other than screen size is returned.

houqp commented:
Yes, that's strange...

```
 $ git submodule status                                                                                   
 01a26bed05c8f5ed36e015d16cc54e52dddcccb7 djvulibre (debian/3.5.24-9-24-g01a26be)
 a9516b455e811b9f57f2805d6c08b6744f797083 kpvcrlib/crengine (cr3.0.57-15)
 2e18531e4d48f46eba8c12749723ba23417b8920 libk2pdfopt (remotes/origin/HEAD)
 149e0fb8ec6295325e1e448bbe79b7e9bf2285aa luafilesystem (v1.5.0-15-g149e0fb)
 eb6f890ebd01ee4cf4f31a66d8a946bc8dccd885 luajit-2.0 (v2.0.0-rc1)
 05219d086a5136ee19f643cf062bd5c0d3aef5d3 mupdf (1.1-233-g05219d0)
```

houqp commented:
OK, it's working now after `make clean`, though I have no idea why `make cleanthirdparty` is not enough...

chrox commented:
That makes sense. Because koptcontext.o is dependent on libk2pdfopt/koptreflow.h and in recent libk2pdfopt I added a new field called dev_dpi in KOPTContext stucture.

houqp commented:
Ok, cool, merging now.

chrox commented on frontend/ui/reader/readerview.lua:
```diff
@@ -44,6 +44,7 @@ function ReaderView:paintTo(bb, x, y)
 			self.state.zoom,
 			self.state.rotation,
 			self.render_mode)
+		self:recalculate()
```
@houqp I would like to see your comment on this. Because this recalculation is a hack for koptreader which cannot determine page size before the page is rendered. Should we have a more elegant way to update the page size?

houqp replied:
Yes, we need a more elegant way. I will look into it this weekend. Feel free to change the UI framework if you come up with any good idea.

houqp replied:
how about call `page:reflow(kc, render_mode)` in `getPageDimensions(pageno, zoom, rotation)` and put `kc` in cache, so `KoptDocument:renderPage` can get it from cache?

chrox replied:
Well. It is better than the recalculation hack. And I will implement it in this way.

houqp replied:
I removed the recalculate call in my latest commit: https://github.com/houqp/kindlepdfviewer/commit/ae1c489a0fb94317bef1fa5a60b4d320af4f8ea6

So we might have a conflict for your next RP, but we can handle it manually later.

BTW, if you have better idea for the implementation, just ignore my suggestion ;P
